### PR TITLE
Sing history

### PR DESCRIPTION
**Pull request by ab3327**
Created sing history, tracks sing time, sing count, quarks, golden quarks, total wow cube blessings,/tess/plat/hyper, c15 score, and quark hept amount.

### CODE REVIEW

File: src/History.ts [hunk 1]
```diff
@@ -62,13 +62,27 @@ export type ResetHistoryEntryAscend = ResetHistoryEntryBase & {
     kind: 'ascend'
 }
 
+export type ResetHistoryEntrySingularity = ResetHistoryEntryBase & {
```
Comment (KhafraDev) [inline src/History.ts:65]: use an interface

File: src/Reset.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { player, clearInt, interval, format, blankSave } from './Synergism';
+import { player, clearInt, interval, format, blankSave} from './Synergism';
```
Comment (KhafraDev) [inline src/Reset.ts:1]: undo this change

File: src/Reset.ts [hunk 3]
```diff
@@ -35,6 +36,8 @@ import { calculateSingularityDebuff } from './singularity';
 import { updateCubeUpgradeBG } from './Cubes';
 import { calculateTessBuildingsInBudget, buyTesseractBuilding } from './Buy'
 import type { TesseractBuildings } from './Buy';
+//import {cubeArray} from './UpdateVisuals';
+import {sumContents} from './Utility';
```
Comment (KhafraDev) [inline src/Reset.ts:40]:
```suggestion
import { sumContents } from './Utility';
```

File: src/Reset.ts [hunk 4]
```diff
@@ -873,6 +876,37 @@ export const singularity = async (): Promise<void> => {
...
+        quarks: player.quarksThisSingularity,
+        c15Score: player.challenge15Exponent,
+        goldenQuarks: calculateGoldenQuarkGain(),
+        wowTribs: sumContents(cubeArray.slice(1) as number[]),
+        tessTribs: sumContents(tesseractArray.slice(1) as number[]),
+        hyperTribs: sumContents(hypercubeArray.slice(1) as number[]),
+        platTribs: sumContents(platonicArray.slice(1)),
+        quarkHept: player.hepteractCrafts.quark.BAL,
+        kind: 'singularity'
+    }
```
Comment (KhafraDev) [inline src/Reset.ts:908]:
```suggestion
    // get total cube blessings for history
    const cubeArray = [player.cubeBlessings.accelerator, player.cubeBlessings.multiplier, player.cubeBlessings.offering,
        player.cubeBlessings.runeExp, player.cubeBlessings.obtainium, player.cubeBlessings.antSpeed, player.cubeBlessings.antSacrifice,
        player.cubeBlessings.antELO, player.cubeBlessings.talismanBonus, player.cubeBlessings.globalSpeed]
    const tesseractArray = [player.tesseractBlessings.accelerator, player.tesseractBlessings.multiplier,
        player.tesseractBlessings.offering, player.tesseractBlessings.runeExp, player.tesseractBlessings.obtainium,
        player.tesseractBlessings.antSpeed, player.tesseractBlessings.antSacrifice, player.tesseractBlessings.antELO,
        player.tesseractBlessings.talismanBonus, player.tesseractBlessings.globalSpeed]
    const hypercubeArray = [player.hypercubeBlessings.accelerator, player.hypercubeBlessings.multiplier,
        player.hypercubeBlessings.offering, player.hypercubeBlessings.runeExp, player.hypercubeBlessings.obtainium,
        player.hypercubeBlessings.antSpeed, player.hypercubeBlessings.antSacrifice, player.hypercubeBlessings.antELO,
        player.hypercubeBlessings.talismanBonus, player.hypercubeBlessings.globalSpeed]
    const platonicArray = [player.platonicBlessings.cubes, player.platonicBlessings.tesseracts, player.platonicBlessings.hypercubes,
        player.platonicBlessings.platonics, player.platonicBlessings.hypercubeBonus, player.platonicBlessings.taxes,
        player.platonicBlessings.scoreBonus, player.platonicBlessings.globalSpeed]
    // Update sing history
    const historyEntry: ResetHistoryEntrySingularity = {
        seconds: player.singularityCounter,
        date: Date.now(),
        singularityCount: player.singularityCount,
        quarks: player.quarksThisSingularity,
        c15Score: player.challenge15Exponent,
        goldenQuarks: calculateGoldenQuarkGain(),
        wowTribs: sumContents(cubeArray),
        tessTribs: sumContents(tesseractArray),
        hyperTribs: sumContents(hypercubeArray),
        platTribs: sumContents(platonicArray),
        quarkHept: player.hepteractCrafts.quark.BAL,
        kind: 'singularity'
    }
```

File: src/Reset.ts [hunk 5]
```diff
@@ -889,6 +923,7 @@ export const singularity = async (): Promise<void> => {
     toggleSubTab(9, 0); // set 'corruption main'
     toggleSubTab(-1, 0); // set 'statistics main'
 
+    hold.history['singularity'] = player.history['singularity'];
```
Comment (KhafraDev) [inline src/Reset.ts:926]: use [dot notation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_accessors#dot_notation)

File: src/Synergism.ts [hunk 6]
```diff
@@ -626,7 +626,7 @@ export const player: Player = {
     corruptionShowStats: true,
 
     constantUpgrades: [null, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
-    history: { ants: [], ascend: [], reset: [] },
+    history: { ants: [], ascend: [], reset: [], singularity: []},
```
Comment (KhafraDev) [inline src/Synergism.ts:629]:
```suggestion
    history: { ants: [], ascend: [], reset: [], singularity: [] },
```

File: Synergism.css [hunk 7]
```diff
@@ -2597,6 +2606,10 @@ p[id$="BlessingsTotal"] {
     min-width: 1212px;
 }
 
+#historySingularity .historyTableWrap {
```
Comment (KhafraDev) [inline Synergism.css:2609]: re-use the rule as mentioned above

**Review by KhafraDev — CHANGES_REQUESTED** [review]

File: src/Reset.ts [hunk 8]
```diff
@@ -873,6 +876,37 @@ export const singularity = async (): Promise<void> => {
...
     }
 
+    // get total cube blessings for history
+    const cubeArray = [null, player.cubeBlessings.accelerator, player.cubeBlessings.multiplier, player.cubeBlessings.offering,
+        player.cubeBlessings.runeExp, player.cubeBlessings.obtainium, player.cubeBlessings.antSpeed, player.cubeBlessings.antSacrifice,
+        player.cubeBlessings.antELO, player.cubeBlessings.talismanBonus, player.cubeBlessings.globalSpeed]
+    const tesseractArray = [null, player.tesseractBlessings.accelerator, player.tesseractBlessings.multiplier,
+        player.tesseractBlessings.offering, player.tesseractBlessings.runeExp, player.tesseractBlessings.obtainium,
+        player.tesseractBlessings.antSpeed, player.tesseractBlessings.antSacrifice, player.tesseractBlessings.antELO,
+        player.tesseractBlessings.talismanBonus, player.tesseractBlessings.globalSpeed]
```
Comment (KhafraDev) [inline src/Reset.ts:886]:
```suggestion
    const tesseractArray = Object.values(player.tesseractBlessings);
```

File: src/Reset.ts [hunk 9]
```diff
@@ -873,6 +876,37 @@ export const singularity = async (): Promise<void> => {
...
+    const hypercubeArray = [null, player.hypercubeBlessings.accelerator, player.hypercubeBlessings.multiplier,
+        player.hypercubeBlessings.offering, player.hypercubeBlessings.runeExp, player.hypercubeBlessings.obtainium,
+        player.hypercubeBlessings.antSpeed, player.hypercubeBlessings.antSacrifice, player.hypercubeBlessings.antELO,
+        player.hypercubeBlessings.talismanBonus, player.hypercubeBlessings.globalSpeed]
```
Comment (KhafraDev) [inline src/Reset.ts:890]:
```suggestion
    const hypercubeArray = Object.values(player.hypercubeBlessings);
```

File: src/Reset.ts [hunk 10]
```diff
@@ -873,6 +876,37 @@ export const singularity = async (): Promise<void> => {
...
+    const platonicArray = [player.platonicBlessings.cubes, player.platonicBlessings.tesseracts, player.platonicBlessings.hypercubes,
+        player.platonicBlessings.platonics, player.platonicBlessings.hypercubeBonus, player.platonicBlessings.taxes,
+        player.platonicBlessings.scoreBonus, player.platonicBlessings.globalSpeed]
```
Comment (KhafraDev) [inline src/Reset.ts:893]:
```suggestion
    const platonicArray = Object.values(player.platonicBlessings);
```

**Review by KhafraDev — CHANGES_REQUESTED** [review]

File: src/History.ts (hunk 1 above)
Comment (ab3327) [reply src/History.ts:65]: how would i use the ResetHistoryEntryBase interface in an interface? All the other ones did it this way

Comment (KhafraDev) [reply src/History.ts:65]: https://www.typescriptlang.org/docs/handbook/interfaces.html#extending-interfaces

**Comment by ab3327** [comment]
I will get to the rest of these in the morning, having issues rn but probably bc I'm tired

File: Synergism.css (hunk 7 above)
Comment (ab3327) [reply Synergism.css:2609]: Do you mean the dot-notation rule so change it to #historySingularity.historyTableWrap {
Or is there another rule I'm not noticing?

Comment (awWhy) [reply Synergism.css]: He means stuff like this
image
Can be done like this
#historySingularity,
#historyAscend {
    margin: 0 auto;
}
